action/cleanup.go

```diff
@@ -0,0 +1,81 @@
+package action
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"sync"
+
+	ctx "github.com/hortonworks/cloud-haunter/context"
+	"github.com/hortonworks/cloud-haunter/types"
+	log "github.com/sirupsen/logrus"
+)
+
+var defaultRetentionDays = 90
+
+type cleanupAction struct {
+	retentionDays int
+}
+
+func init() {
+	retentionEnv := os.Getenv("RETENTION_DAYS")
+	var retentionDays int
+	if len(retentionEnv) > 0 {
+		parsed, err := strconv.Atoi(retentionEnv)
+		if err != nil {
+			log.Fatalf("[CLEANUP] Failed to parse RETENTION_DAYS, err: %s", err)
+		}
+		retentionDays = parsed
+	} else {
+		retentionDays = defaultRetentionDays
+	}
+	ctx.Actions[types.CleanupAction] = cleanupAction{retentionDays}
+}
+
+func (a cleanupAction) Execute(op types.OpType, filters []types.FilterType, items []types.CloudItem) {
+	wg := sync.WaitGroup{}
+	wg.Add(len(ctx.CloudProviders))
+	for t, p := range ctx.CloudProviders {
+		go func(cType types.CloudType, provider types.CloudProvider) {
+			defer wg.Done()
+
+			var cloudItems []*types.CloudItem
+			for _, item := range items {
+				if item.GetCloudType() == cType {
+					i := item
+					cloudItems = append(cloudItems, &i)
+				}
+			}
+
+			if len(cloudItems) > 0 {
+				log.Infof("[CLEANUP] Cleaning up %d items on %s: %s", len(cloudItems), cType, items)
+				var errors []error
+
+				item := *cloudItems[0]
+				switch t := item.GetItem().(type) {
+				case types.Storage:
+					errors = a.cleanupStorages(provider, cloudItems)
+				default:
+					panic(fmt.Sprintf("[CLEANUP] Operation on type %T is not allowed", t))
```

Comment:
Does it makes sense to do this validation earlier?

Reply:
From the interface perspective, it could happen that we only have a single non-storage item, so I think it is fine, we do it similarly in other actions like termination